File: Problems/Algorithms/931. Minimum Falling Path Sum/min_falling_path_sum.py
#!/usr/bin/env python3

from typing import List


class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        m, n = len(matrix), len(matrix[0])
        dp = [[0 for _ in range(n)] for _ in range(m)]

        for j in range(n):
            dp[0][j] = matrix[0][j]
        
        for i in range(1, m):
            for j in range(n):
                left = float('inf') if j == 0 else dp[i-1][j-1]
                right = float('inf') if j == n-1 else dp[i-1][j+1]
                mid = dp[i-1][j]
                dp[i][j] = matrix[i][j] + min(mid, min(left, right))
        
        res = dp[m-1][0]
        for j in range(1, n):
            res = min(res, dp[m-1][j])
        return res

File: Problems/Algorithms/931. Minimum Falling Path Sum/min-falling-path-sum.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        int n = matrix.size();
        int m = matrix[0].size();
        vector<vector<int>> dp(n, vector<int>(m, 0));

        for (int j = 0; j < m; j++) {
            dp[0][j] = matrix[0][j];
        }

        for (int i = 1; i < n; i++) {
            for (int j = 0; j < m; j++) {
                int left = 1e9, right = 1e9;
                int mid = matrix[i][j] + dp[i-1][j];

                if (j-1 >= 0) {
                    left = matrix[i][j] + dp[i-1][j-1];
                }
                if (j+1 < m) {
                    right = matrix[i][j] + dp[i-1][j+1];
                }

                dp[i][j] = min(mid, min(left, right));
            }
        }
        
        int ans = dp[n-1][0];
        for (int j = 1; j < m; j++) {
            ans = min(ans, dp[n-1][j]);
        }

        return ans;
    }
};
